File: character_logic/dracle.py
"""Dracle battle and AI hooks."""

from __future__ import annotations

from typing import Any

from .common import floor_int


CHARACTER_ID = "dracle"
DRAGON = "혁룡"
DAMAGE_RECORD = "혁룡 피해 기록"
AWAKENING = "혁룡 각성"
SCALE_GUARD = "용의 비늘"
BLOODLUST = "용혈의 투지"
HIDDEN_COUNTERS = {DAMAGE_RECORD, AWAKENING, SCALE_GUARD, BLOODLUST}

BASE_MAX_DRAGON = 10
AWAKENED_MAX_DRAGON = 15
AWAKENING_GAIN = 5
AWAKENING_TURNS = 4
AWAKENING_READY_MP = 60
DAMAGE_RECORD_PER_STACK = 15
BREATH_DAMAGE_MULTIPLIER = 1.3
BREATH_MP_DRAIN_RATIO = 0.3
CLAW_ATK_MULTIPLIER = 1.2
CLAW_ATK_TURNS = 3


def init_unique_state(fighter: Any, unique_names: set[str]) -> None:
    if DRAGON in unique_names:
        fighter.counters[DRAGON] = 0
        fighter.counters[DAMAGE_RECORD] = 0


def counter_state_text(fighter: Any, name: str, value: Any) -> str | None:
    if name == DRAGON:
        return f"{DRAGON} {int(value)}/{_max_dragon(fighter)}"
    return None


def extra_state_parts(battle: Any, fighter: Any) -> list[str]:
    parts: list[str] = []
    record = int(fighter.counters.get(DAMAGE_RECORD, 0))
    if record > 0 or _dragon(fighter) < _max_dragon(fighter):
        parts.append(f"{DAMAGE_RECORD} {record}/{DAMAGE_RECORD_PER_STACK}")
    if int(fighter.counters.get(AWAKENING, 0)) > 0:
        parts.append(f"{AWAKENING} {fighter.counters[AWAKENING]}턴")
    if int(fighter.counters.get(SCALE_GUARD, 0)) > 0:
        parts.append(f"{SCALE_GUARD}: 공격 피해 -{_scale_reduction(fighter)}")
    if int(fighter.counters.get(BLOODLUST, 0)) > 0:
        parts.append(f"{BLOODLUST}: 공격 피해 x1.1")
    return parts


def reset_turn_flags(battle: Any, fighter: Any) -> None:
    fighter.counters.pop(SCALE_GUARD, None)


def counter_resource_value(fighter: Any, name: str, raw: Any) -> float | None:
    if name == DRAGON and isinstance(raw, int):
        return raw * 115
    if name == DAMAGE_RECORD and isinstance(raw, int):
        return raw * 8
    if name == AWAKENING:
        return 520 if int(raw or 0) > 0 else 0.0
    if name == BLOODLUST:
        return 190 if int(raw or 0) > 0 else 0.0
    return None


def on_action_start(battle: Any, choice: Any) -> bool:
    actor = choice.actor
    action = choice.action
    if action.is_skill(CHARACTER_ID, 0) and _dragon(actor) >= 1:
        battle.restore_mp(actor, 3, action.name)
    return False


def apply_condition_effects(battle: Any, choice: Any) -> bool | None:
    actor = choice.actor
    action = choice.action
    dragon = _dragon(actor)
    if action.is_skill(CHARACTER_ID, 0) and dragon >= 4:
        bonus = floor_int(dragon * 1.5)
        choice.power = (choice.power or 0) + bonus
        print(f"{DRAGON} {dragon}/{_max_dragon(actor)} 기준으로 위력이 {bonus} 증가했다.")
    if action.is_skill(CHARACTER_ID, 1) and dragon < 1:
        return False
    if action.is_skill(CHARACTER_ID, 3) and (dragon < 5 or _awakening_active(actor)):
        return False
    return None


def estimated_power(battle: Any, actor: Any, target: Any, action: Any, power: int) -> int:
    dragon = _dragon(actor)
    if action.is_skill(CHARACTER_ID, 0) and dragon >= 4:
        power += floor_int(dragon * 1.5)
    return power


def attack_damage_multipliers(battle: Any, choice: Any) -> list[float]:
    actor = choice.actor
    action = choice.action
    dragon = _dragon(actor)
    multipliers: list[float] = []
    if int(actor.counters.get(BLOODLUST, 0)) > 0:
        multipliers.append(1.1)
    if action.is_skill(CHARACTER_ID, 0) and dragon >= 7:
        multipliers.append(BREATH_DAMAGE_MULTIPLIER)
    if action.is_skill(CHARACTER_ID, 2):
        chance = _claw_double_chance(actor)
        roll = battle.roll("용의 발톱 피해 증폭")
        print(f"용의 발톱 피해 증폭 {chance}% / 판정값 {roll:.2f}")
        if roll < chance:
            multipliers.append(2.0)
    return multipliers


def estimated_damage_multipliers(battle: Any, actor: Any, target: Any, action: Any) -> list[float]:
    dragon = _dragon(actor)
    multipliers: list[float] = []
    if int(actor.counters.get(BLOODLUST, 0)) > 0:
        multipliers.append(1.1)
    if action.is_skill(CHARACTER_ID, 0) and dragon >= 7:
        multipliers.append(BREATH_DAMAGE_MULTIPLIER)
    if action.is_skill(CHARACTER_ID, 2):
        multipliers.append(1 + _claw_double_chance(actor) / 100)
    return multipliers


def modify_attack_damage(battle: Any, choice: Any, target: Any, damage: int) -> int:
    if int(target.counters.get(SCALE_GUARD, 0)) <= 0:
        return damage
    reduction = _scale_reduction(target)
    if reduction <= 0:
        return damage
    reduced = max(1, damage - reduction)
    if reduced < damage:
        print(f"{SCALE_GUARD}로 공격 피해가 {damage} → {reduced}로 감소했다.")
    return reduced


def on_hit_after_defense(battle: Any, choice: Any, total_damage: int) -> None:
    actor = choice.actor
    target = battle.opponent(actor)
    action = choice.action
    _record_dragon_damage(battle, actor, total_damage)
    if action.is_skill(CHARACTER_ID, 0) and _dragon(actor) >= 10:
        battle.reduce_mp(target, floor_int(total_damage * BREATH_MP_DRAIN_RATIO), action.name)
    if action.is_skill(CHARACTER_ID, 2):
        battle.add_stat_effect(actor, "atk", CLAW_ATK_MULTIPLIER, CLAW_ATK_TURNS, action.name)


def apply_non_attack_effects(battle: Any, choice: Any) -> bool:
    actor = choice.actor
    action = choice.action
    if action.is_skill(CHARACTER_ID, 1):
        actor.counters[SCALE_GUARD] = 1
        print(f"이번 턴 동안 공격 피해가 {_scale_reduction(actor)} 감소한다.")
        return True
    if action.is_skill(CHARACTER_ID, 3):
        actor.counters[AWAKENING] = AWAKENING_TURNS
        battle.add_counter(actor, DRAGON, AWAKENING_GAIN, max_value=_max_dragon(actor))
        print(f"{AWAKENING}으로 {DRAGON} 최대 중첩이 {_max_dragon(actor)}가 되었다.")
        _convert_damage_record(battle, actor)
        return True
    return False


def finish_action(battle: Any, choice: Any, success: bool, hit: bool, miss_not_failure: bool) -> None:
    if choice.action.is_common_action("normal_attack") and success and hit:
        choice.actor.counters[BLOODLUST] = 2
        print(f"다음 턴 동안 {BLOODLUST}로 공격 피해가 1.1배가 된다.")


def decrement_counters(fighter: Any) -> None:
    if int(fighter.counters.get(BLOODLUST, 0)) > 0:
        remaining = int(fighter.counters.get(BLOODLUST, 0)) - 1
        if remaining <= 0:
            fighter.counters.pop(BLOODLUST, None)
        else:
            fighter.counters[BLOODLUST] = remaining
    if int(fighter.counters.get(AWAKENING, 0)) <= 0:
        return
    remaining = int(fighter.counters.get(AWAKENING, 0))
    if remaining <= 1:
        fighter.counters.pop(AWAKENING, None)
        before = _dragon(fighter)
        fighter.counters[DRAGON] = min(BASE_MAX_DRAGON, max(0, before - AWAKENING_GAIN))
        print(f"{AWAKENING} 종료: {DRAGON} {before}/{AWAKENED_MAX_DRAGON} → {fighter.counters[DRAGON]}/{BASE_MAX_DRAGON}")
    else:
        fighter.counters[AWAKENING] = remaining - 1


def setup_value(battle: Any, actor: Any, target: Any, action: Any) -> float:
    dragon = _dragon(actor)
    incoming = battle.estimate_best_incoming_damage(target, actor)
    if action.is_skill(CHARACTER_ID, 1):
        if dragon < 1:
            return 0.0
        reduction = _scale_reduction(actor)
        if incoming < actor.hp * 0.45:
            return 0.0
        value = min(incoming, reduction) * 55
        if incoming >= actor.hp:
            value += 1200
        elif incoming >= actor.hp * 0.45:
            value += 520
        return value
    if not action.is_skill(CHARACTER_ID, 3):
        return 0.0
    if dragon < 5 or _awakening_active(actor):
        return 0.0
    if actor.mp < _awakening_ready_mp(action):
        return -2600 - max(0, _awakening_ready_mp(action) - actor.mp) * 22
    value = 2100 + dragon * 190 + int(actor.counters.get(DAMAGE_RECORD, 0)) * 24
    if actor.mp >= _awakening_ready_mp(action) + 10:
        value += 420
    if incoming >= actor.hp:
        value -= 1500
    return value


def ai_score(
    battle: Any,
    actor: Any,
    target: Any,
    action: Any,
    expected_damage: float,
    hit_rate: float,
) -> float:
    value = 0.0
    dragon = _dragon(actor)
    incoming = battle.estimate_best_incoming_damage(target, actor)
    awakening_action = actor.skill_by_key(battle.skill_key(CHARACTER_ID, 3))
    ready_mp = _awakening_ready_mp(awakening_action) if awakening_action else AWAKENING_READY_MP
    wants_awakening = dragon >= 5 and not _awakening_active(actor)

    if action.is_common_action("normal_attack"):
        value += 220 + _dragon_gain_value(actor, expected_damage * hit_rate)
        if actor.mp < 35:
            value += 180
    elif action.is_common_action("meditation"):
        if wants_awakening and actor.mp < ready_mp:
            value += 1100 + (ready_mp - actor.mp) * 28
        elif actor.mp < 45:
            value += 420
        if actor.mp >= 88:
            value -= 650

    if action.is_skill(CHARACTER_ID, 0):
        value += 260 + _dragon_gain_value(actor, expected_damage * hit_rate)
        if dragon < 5:
            value += max(0, 5 - dragon) * 120
        if dragon >= 1:
            value += 210
        if dragon >= 4:
            value += dragon * 65
        if dragon >= 7:
            value += expected_damage * 0.75
        if dragon >= 10 and target.mp > 0:
            value += min(target.mp, floor_int(expected_damage * BREATH_MP_DRAIN_RATIO)) * 38

    elif action.is_skill(CHARACTER_ID, 1):
        if dragon < 1:
            value -= 3200
        else:
            reduction = _scale_reduction(actor)
            prevented = min(incoming, reduction)
            value += prevented * 70 + dragon * 20
            if incoming >= actor.hp:
                value += 1800
            elif incoming >= actor.hp * 0.55:
                value += 820
            else:
                value -= 520
            if incoming <= reduction * 0.45:
                value -= 260
            if wants_awakening and actor.mp < ready_mp:
                value -= 650

    elif action.is_skill(CHARACTER_ID, 2):
        value += _dragon_gain_value(actor, expected_damage * hit_rate)
        chance = _claw_double_chance(actor)
        value += chance * 12
        value += (180 if _has_stat_effect(actor, "atk", action.name) else 520) * hit_rate
        if expected_damage >= target.hp:
            value += 3600
        elif dragon >= 5 and not _awakening_active(actor):
            cost = battle.effective_cost(actor, action)
            if actor.mp - cost + 10 < ready_mp and expected_damage < target.hp * 0.65:
                value -= 780
        if dragon >= 8:
            value += 520

    elif action.is_skill(CHARACTER_ID, 3):
        if dragon < 5:
            value -= 4800
        elif _awakening_active(actor):
            value -= 3600
        elif actor.mp < ready_mp:
            value -= 3600 + (ready_mp - actor.mp) * 32
        else:
            value += 3200 + dragon * 260
            value += int(actor.counters.get(DAMAGE_RECORD, 0)) * 34
            if actor.mp >= ready_mp + 10:
                value += 620
            if incoming >= actor.hp:
                value -= 1800

    return value


def would_condition_fail(battle: Any, actor: Any, target: Any, action: Any) -> bool | None:
    if action.is_skill(CHARACTER_ID, 1):
        return _dragon(actor) < 1
    if action.is_skill(CHARACTER_ID, 3):
        return _dragon(actor) < 5 or _awakening_active(actor)
    return None


def _dragon(fighter: Any) -> int:
    return int(fighter.counters.get(DRAGON, 0))


def _max_dragon(fighter: Any) -> int:
    return AWAKENED_MAX_DRAGON if _awakening_active(fighter) else BASE_MAX_DRAGON


def _awakening_active(fighter: Any) -> bool:
    return int(fighter.counters.get(AWAKENING, 0)) > 0


def _scale_reduction(fighter: Any) -> int:
    return _dragon(fighter) * 4


def _claw_double_chance(fighter: Any) -> int:
    return min(100, _dragon(fighter) * 10)


def _awakening_ready_mp(action: Any | None) -> int:
    cost = int(getattr(action, "mp", 43) if action is not None else 43)
    return max(AWAKENING_READY_MP, cost + 17)


def _has_stat_effect(fighter: Any, stat: str, source: str) -> bool:
    return any(
        getattr(effect, "stat", None) == stat
        and getattr(effect, "source", None) == source
        and int(getattr(effect, "remaining", 0)) > 0
        for effect in getattr(fighter, "stat_effects", [])
    )


def _dragon_gain_value(actor: Any, expected_damage: float) -> float:
    if _dragon(actor) >= _max_dragon(actor):
        return 0.0
    record = int(actor.counters.get(DAMAGE_RECORD, 0))
    projected = record + max(0, expected_damage)
    gained = min(_max_dragon(actor) - _dragon(actor), floor_int(projected / DAMAGE_RECORD_PER_STACK))
    return gained * 520 + min(DAMAGE_RECORD_PER_STACK - 1, projected % DAMAGE_RECORD_PER_STACK) * 18


def _record_dragon_damage(battle: Any, actor: Any, total_damage: int) -> None:
    if total_damage <= 0 or _dragon(actor) >= _max_dragon(actor):
        return
    before = int(actor.counters.get(DAMAGE_RECORD, 0))
    actor.counters[DAMAGE_RECORD] = before + total_damage
    print(f"{DAMAGE_RECORD} {before} → {actor.counters[DAMAGE_RECORD]}")
    _convert_damage_record(battle, actor)


def _convert_damage_record(battle: Any, actor: Any) -> None:
    record = int(actor.counters.get(DAMAGE_RECORD, 0))
    current = _dragon(actor)
    max_dragon = _max_dragon(actor)
    gained = min(max_dragon - current, record // DAMAGE_RECORD_PER_STACK)
    if gained <= 0:
        return
    actor.counters[DAMAGE_RECORD] = record - gained * DAMAGE_RECORD_PER_STACK
    actor.counters[DRAGON] = current + gained
    print(f"{actor.name}의 {DRAGON} {current}/{max_dragon} → {actor.counters[DRAGON]}/{max_dragon}")

File: web/battle-engine/character-logic/dracle.js
"use strict";

const CHARACTER_ID = "dracle";
const DRAGON = "혁룡";
const DAMAGE_RECORD = "혁룡 피해 기록";
const AWAKENING = "혁룡 각성";
const SCALE_GUARD = "용의 비늘";
const BLOODLUST = "용혈의 투지";
const BASE_MAX_DRAGON = 10;
const AWAKENED_MAX_DRAGON = 15;

function floorInt(value) { return Math.floor(value); }
function dragon(fighter) { return Number(fighter.counters[DRAGON] || 0); }
function awakeningActive(fighter) { return Number(fighter.counters[AWAKENING] || 0) > 0; }
function maxDragon(fighter) { return awakeningActive(fighter) ? AWAKENED_MAX_DRAGON : BASE_MAX_DRAGON; }
function scaleReduction(fighter) { return dragon(fighter) * 4; }
function clawChance(fighter) { return Math.min(100, dragon(fighter) * 10); }
function awakeningReadyMp(action) { return Math.max(60, Number(action?.mp ?? 43) + 17); }
function hasStatEffect(fighter, stat, source) {
  return fighter.statEffects.some((effect) => effect.stat === stat && effect.source === source && Number(effect.remaining || 0) > 0);
}
function dragonGainValue(actor, expectedDamage) {
  if (dragon(actor) >= maxDragon(actor)) return 0;
  const record = Number(actor.counters[DAMAGE_RECORD] || 0);
  const projected = record + Math.max(0, expectedDamage);
  const gained = Math.min(maxDragon(actor) - dragon(actor), floorInt(projected / 15));
  return gained * 520 + Math.min(14, projected % 15) * 18;
}
function rawDamage(expectedDamage, hitRate) { return hitRate > 0 ? expectedDamage / hitRate : 0; }

function addCounter(battle, fighter, name, amount, maxValue) {
  const before = Number(fighter.counters[name] || 0);
  fighter.counters[name] = Math.min(maxValue, before + amount);
  battle.logs.push(`${fighter.name}의 ${name} ${before} -> ${fighter.counters[name]}`);
}

function convertDamageRecord(battle, actor) {
  const record = Number(actor.counters[DAMAGE_RECORD] || 0);
  const current = dragon(actor);
  const maximum = maxDragon(actor);
  const gained = Math.min(maximum - current, floorInt(record / 15));
  if (gained <= 0) return;
  actor.counters[DAMAGE_RECORD] = record - gained * 15;
  actor.counters[DRAGON] = current + gained;
  battle.logs.push(`${actor.name}의 ${DRAGON} ${current}/${maximum} -> ${actor.counters[DRAGON]}/${maximum}`);
}

module.exports = {
  hiddenCounters: [DAMAGE_RECORD, AWAKENING, SCALE_GUARD, BLOODLUST],

  initUniqueState(fighter, uniqueNames) {
    if (uniqueNames.has(DRAGON)) {
      fighter.counters[DRAGON] = 0;
      fighter.counters[DAMAGE_RECORD] = 0;
    }
  },

  counterStateText(fighter, name, value) {
    return name === DRAGON ? `${DRAGON} ${Number(value)}/${maxDragon(fighter)}` : null;
  },

  counterResourceValue(fighter, name, raw) {
    if (name === DRAGON && Number.isInteger(raw)) return raw * 115;
    if (name === DAMAGE_RECORD && Number.isInteger(raw)) return raw * 8;
    if (name === AWAKENING) return Number(raw || 0) > 0 ? 520 : 0;
    if (name === BLOODLUST) return Number(raw || 0) > 0 ? 190 : 0;
    return null;
  },

  extraStateParts(_battle, fighter) {
    const parts = [];
    const record = Number(fighter.counters[DAMAGE_RECORD] || 0);
    if (record > 0 || dragon(fighter) < maxDragon(fighter)) parts.push(`${DAMAGE_RECORD} ${record}/15`);
    if (Number(fighter.counters[AWAKENING] || 0) > 0) parts.push(`${AWAKENING} ${fighter.counters[AWAKENING]}턴`);
    if (Number(fighter.counters[SCALE_GUARD] || 0) > 0) parts.push(`${SCALE_GUARD}: 공격 피해 -${scaleReduction(fighter)}`);
    if (Number(fighter.counters[BLOODLUST] || 0) > 0) parts.push(`${BLOODLUST}: 공격 피해 x1.1`);
    return parts;
  },

  resetTurnFlags(_battle, fighter) {
    delete fighter.counters[SCALE_GUARD];
  },

  onActionStart(battle, choice) {
    if (choice.action.isSkill(CHARACTER_ID, 0) && dragon(choice.actor) >= 1) {
      battle.restoreMp(choice.actor, 3, choice.action.name);
    }
    return false;
  },

  applyConditionEffects(battle, choice) {
    const amount = dragon(choice.actor);
    if (choice.action.isSkill(CHARACTER_ID, 0) && amount >= 4) {
      const bonus = floorInt(amount * 1.5);
      choice.power = Number(choice.power || 0) + bonus;
      battle.logs.push(`${DRAGON} ${amount}/${maxDragon(choice.actor)} 기준으로 위력이 ${bonus} 증가했다.`);
    }
    if (choice.action.isSkill(CHARACTER_ID, 1) && amount < 1) return false;
    if (choice.action.isSkill(CHARACTER_ID, 3) && (amount < 5 || awakeningActive(choice.actor))) return false;
    return true;
  },

  attackDamageMultipliers(battle, choice) {
    const actor = choice.actor;
    const values = [];
    if (Number(actor.counters[BLOODLUST] || 0) > 0) values.push(1.1);
    if (choice.action.isSkill(CHARACTER_ID, 0) && dragon(actor) >= 7) values.push(1.3);
    if (choice.action.isSkill(CHARACTER_ID, 2)) {
      const chance = clawChance(actor);
      const roll = battle.roll("용의 발톱 피해 증폭");
      battle.logs.push(`용의 발톱 피해 증폭 ${chance}% / 판정값 ${roll.toFixed(2)}`);
      if (roll < chance) values.push(2);
    }
    return values;
  },

  estimatedDamageMultipliers(_battle, actor, _target, action) {
    const values = [];
    if (Number(actor.counters[BLOODLUST] || 0) > 0) values.push(1.1);
    if (action.isSkill(CHARACTER_ID, 0) && dragon(actor) >= 7) values.push(1.3);
    if (action.isSkill(CHARACTER_ID, 2)) values.push(1 + clawChance(actor) / 100);
    return values;
  },

  modifyAttackDamage(battle, _choice, target, damage) {
    if (Number(target.counters[SCALE_GUARD] || 0) <= 0) return damage;
    const reduced = Math.max(1, damage - scaleReduction(target));
    if (reduced < damage) battle.logs.push(`${SCALE_GUARD}로 공격 피해가 ${damage} -> ${reduced}로 감소했다.`);
    return reduced;
  },

  onHitAfterDefenseAsActor(battle, choice, totalDamage) {
    const actor = choice.actor;
    if (totalDamage > 0 && dragon(actor) < maxDragon(actor)) {
      const before = Number(actor.counters[DAMAGE_RECORD] || 0);
      actor.counters[DAMAGE_RECORD] = before + totalDamage;
      battle.logs.push(`${DAMAGE_RECORD} ${before} -> ${actor.counters[DAMAGE_RECORD]}`);
      convertDamageRecord(battle, actor);
    }
    if (choice.action.isSkill(CHARACTER_ID, 0) && dragon(actor) >= 10) {
      battle.reduceMp(battle.opponent(actor), floorInt(totalDamage * 0.3), choice.action.name);
    }
    if (choice.action.isSkill(CHARACTER_ID, 2)) battle.addStatEffect(actor, "atk", 1.2, 3, choice.action.name);
  },

  applyNonAttackEffects(battle, choice) {
    const actor = choice.actor;
    if (choice.action.isSkill(CHARACTER_ID, 1)) {
      actor.counters[SCALE_GUARD] = 1;
      battle.logs.push(`이번 턴 동안 공격 피해가 ${scaleReduction(actor)} 감소한다.`);
      return true;
    }
    if (choice.action.isSkill(CHARACTER_ID, 3)) {
      actor.counters[AWAKENING] = 4;
      addCounter(battle, actor, DRAGON, 5, maxDragon(actor));
      battle.logs.push(`${AWAKENING}으로 ${DRAGON} 최대 중첩이 ${maxDragon(actor)}가 되었다.`);
      convertDamageRecord(battle, actor);
      return true;
    }
    return false;
  },

  finishAction(battle, choice, success, hit) {
    if (choice.action.isCommonAction("normal_attack") && success && hit) {
      choice.actor.counters[BLOODLUST] = 2;
      battle.logs.push(`다음 턴 동안 ${BLOODLUST}로 공격 피해가 1.1배가 된다.`);
    }
  },

  decrementCounters(fighter) {
    if (Number(fighter.counters[BLOODLUST] || 0) > 0) {
      fighter.counters[BLOODLUST] -= 1;
      if (fighter.counters[BLOODLUST] <= 0) delete fighter.counters[BLOODLUST];
    }
    const remaining = Number(fighter.counters[AWAKENING] || 0);
    if (remaining <= 0) return;
    if (remaining <= 1) {
      delete fighter.counters[AWAKENING];
      const before = dragon(fighter);
      fighter.counters[DRAGON] = Math.min(BASE_MAX_DRAGON, Math.max(0, before - 5));
    } else {
      fighter.counters[AWAKENING] = remaining - 1;
    }
  },

  estimatedPower(_battle, actor, _target, action, power) {
    return action.isSkill(CHARACTER_ID, 0) && dragon(actor) >= 4 ? power + floorInt(dragon(actor) * 1.5) : power;
  },

  setupValue(battle, actor, target, action) {
    const amount = dragon(actor);
    const incoming = battle.estimateBestIncomingDamage(target, actor);
    if (action.isSkill(CHARACTER_ID, 1)) {
      if (amount < 1 || incoming < actor.hp * 0.45) return 0;
      let value = Math.min(incoming, scaleReduction(actor)) * 55;
      if (incoming >= actor.hp) value += 1200;
      else if (incoming >= actor.hp * 0.45) value += 520;
      return value;
    }
    if (!action.isSkill(CHARACTER_ID, 3) || amount < 5 || awakeningActive(actor)) return 0;
    const readyMp = awakeningReadyMp(action);
    if (actor.mp < readyMp) return -2600 - Math.max(0, readyMp - actor.mp) * 22;
    let value = 2100 + amount * 190 + Number(actor.counters[DAMAGE_RECORD] || 0) * 24;
    if (actor.mp >= readyMp + 10) value += 420;
    if (incoming >= actor.hp) value -= 1500;
    return value;
  },

  aiScore(battle, actor, target, action, expectedDamage, hitRate) {
    let value = 0;
    const damage = rawDamage(expectedDamage, hitRate);
    const amount = dragon(actor);
    const incoming = battle.estimateBestIncomingDamage(target, actor);
    const awakeningAction = battle.findActionByInput(actor, `${CHARACTER_ID}:3`);
    const readyMp = awakeningReadyMp(awakeningAction);
    const wantsAwakening = amount >= 5 && !awakeningActive(actor);

    if (action.isCommonAction("normal_attack")) {
      value += 220 + dragonGainValue(actor, damage * hitRate);
      if (actor.mp < 35) value += 180;
    } else if (action.isCommonAction("meditation")) {
      if (wantsAwakening && actor.mp < readyMp) value += 1100 + (readyMp - actor.mp) * 28;
      else if (actor.mp < 45) value += 420;
      if (actor.mp >= 88) value -= 650;
    }

    if (action.isSkill(CHARACTER_ID, 0)) {
      value += 260 + dragonGainValue(actor, damage * hitRate);
      if (amount < 5) value += Math.max(0, 5 - amount) * 120;
      if (amount >= 1) value += 210;
      if (amount >= 4) value += amount * 65;
      if (amount >= 7) value += damage * 0.75;
      if (amount >= 10 && target.mp > 0) value += Math.min(target.mp, floorInt(damage * 0.3)) * 38;
    } else if (action.isSkill(CHARACTER_ID, 1)) {
      if (amount < 1) value -= 3200;
      else {
        const reduction = scaleReduction(actor);
        value += Math.min(incoming, reduction) * 70 + amount * 20;
        if (incoming >= actor.hp) value += 1800;
        else if (incoming >= actor.hp * 0.55) value += 820;
        else value -= 520;
        if (incoming <= reduction * 0.45) value -= 260;
        if (wantsAwakening && actor.mp < readyMp) value -= 650;
      }
    } else if (action.isSkill(CHARACTER_ID, 2)) {
      value += dragonGainValue(actor, damage * hitRate);
      value += clawChance(actor) * 12;
      value += (hasStatEffect(actor, "atk", action.name) ? 180 : 520) * hitRate;
      if (damage >= target.hp) value += 3600;
      else if (amount >= 5 && !awakeningActive(actor)) {
        const cost = battle.effectiveCost(actor, action);
        if (actor.mp - cost + battle.turnEndMpRecovery(actor) < readyMp && damage < target.hp * 0.65) value -= 780;
      }
      if (amount >= 8) value += 520;
    } else if (action.isSkill(CHARACTER_ID, 3)) {
      if (amount < 5) value -= 4800;
      else if (awakeningActive(actor)) value -= 3600;
      else if (actor.mp < readyMp) value -= 3600 + (readyMp - actor.mp) * 32;
      else {
        value += 3200 + amount * 260 + Number(actor.counters[DAMAGE_RECORD] || 0) * 34;
        if (actor.mp >= readyMp + 10) value += 620;
        if (incoming >= actor.hp) value -= 1800;
      }
    }
    return value;
  },

  wouldConditionFail(_battle, actor, _target, action) {
    if (action.isSkill(CHARACTER_ID, 1)) return dragon(actor) < 1;
    if (action.isSkill(CHARACTER_ID, 3)) return dragon(actor) < 5 || awakeningActive(actor);
    return null;
  },
};
